src/core/geometry/qgscurvepolygon.cpp

```diff
@@ -435,6 +442,52 @@ QgsAbstractGeometry *QgsCurvePolygon::boundary() const
   }
 }
 
+QgsCurvePolygon *QgsCurvePolygon::asGridified( double hSpacing, double vSpacing, double dSpacing, double mSpacing,
+    double tolerance, SegmentationToleranceType toleranceType ) const
+{
+  if ( !mExteriorRing )
+    return nullptr;
+
+  QgsPolygonV2 *polygon;
+  if ( QgsWkbTypes::flatType( mWkbType ) == QgsWkbTypes::Triangle ||  QgsWkbTypes::flatType( mWkbType ) == QgsWkbTypes::Polygon )
+  {
+    polygon = static_cast< QgsPolygonV2 const *>( this )->newSameGeometry();
+  }
+  else
+  {
+    polygon = new QgsPolygonV2();
+    polygon->mWkbType = QgsWkbTypes::zmType( QgsWkbTypes::Polygon, QgsWkbTypes::hasZ( mWkbType ), QgsWkbTypes::hasM( mWkbType ) );
+  }
+
+  // exterior ring
+  QgsCurve *exterior = mExteriorRing->asGridified( hSpacing, vSpacing, dSpacing, mSpacing, tolerance, toleranceType );
+
+  if ( !exterior )
+  {
+    delete polygon;
+    return nullptr;
+  }
+
+  polygon->mExteriorRing = exterior;
+
+  //interior rings
+  for ( QgsCurve *interior : mInteriorRings )
+  {
+    if ( !interior )
+      continue;
+
+    QgsCurve *gridifiedInterior = interior->asGridified( hSpacing, vSpacing, dSpacing, mSpacing, tolerance, toleranceType );
+
+    if ( !gridifiedInterior )
+      continue;
+
+    polygon->mInteriorRings.append( gridifiedInterior );
```

Comment:
Again, use `polygon->addInteriorRing( gridifiedInterior )` instead.

Reply:
Similar to above, this case it only does a lot of conditionals which we already know the result of, to en up doing an append.